### PR TITLE
vm snapshot test case

### PR DESCRIPTION
**Pull request by parikshithb**

### CODE REVIEW

File: tests/functional/workloads/cnv/test_vm_cloning_ops.py [hunk 1]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
             ), f"Failed: MD5 comparison between source {vm_obj.name} and cloned {clone_obj.name} VMs"
             run_dd_io(vm_obj=clone_obj, file_path=file_paths[1])
             clone_obj.stop()
+
+    @workloads
+    @pytest.mark.polarion_id("OCS-6299")
+    def test_vm_snapshot_ops(
```
Comment (avd-sagare) [inline tests/functional/workloads/cnv/test_vm_cloning_ops.py:68]: Need to change test script accordingly as it has snapshot tc as well

File: tests/functional/workloads/cnv/test_vm_cloning_ops.py [hunk 2]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        file_paths = ["/file.txt", "/new_file.txt"]
+        # TODO: Add multi_cnv fixture to configure VMs based on specifications
+        vm_obj = cnv_workload(
+            volume_interface=constants.VM_VOLUME_PVC,
+            source_url=constants.CNV_FEDORA_SOURCE,
+        )[0]
+        # Writing IO on source VM
+        source_csum = run_dd_io(vm_obj=vm_obj, file_path=file_paths[0], verify=True)
+        # Stopping VM before taking snapshot of the VM PVC
+        vm_obj.stop()
```
Comment (avd-sagare) [inline tests/functional/workloads/cnv/test_vm_cloning_ops.py:93]: As per doc I guess no need to stop VM.
https://docs.redhat.com/en/documentation/red_hat_openshift_data_foundation/4.17/html/managing_and_allocating_storage_resources/managing-persistent-volume-claims_rhodf#expanding-persistent-volume-claims_rhodf

File: tests/functional/workloads/cnv/test_vm_cloning_ops.py [hunk 3]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        res_snap_obj = snapshot_restore_factory(
+            snapshot_obj=snap_obj,
+            storageclass=vm_obj.sc_name,
+            size=vm_obj.pvc_size,
+            volume_mode=snap_obj.parent_volume_mode,
+            access_mode=vm_obj.pvc_access_mode,
+            status=constants.STATUS_BOUND,
+            timeout=300,
+        )
+        # Create new PVC using the restore PVC
```
Comment (avd-sagare) [inline tests/functional/workloads/cnv/test_vm_cloning_ops.py:107]: I guess  "new VM" instaed ov "new PVC"

Comment (parikshithb) [reply tests/functional/workloads/cnv/test_vm_cloning_ops.py]: done

File: tests/functional/workloads/cnv/test_vm_cloning_ops.py (hunk 1 above)
Comment (parikshithb) [reply tests/functional/workloads/cnv/test_vm_cloning_ops.py:68]: ack, renamed it

**Review by avd-sagare — APPROVED** [review]

**Review by hnallurv — APPROVED** [review]

File: tests/conftest.py [hunk 4]
```diff
@@ -7088,7 +7091,8 @@ def teardown():
         Cleans up the CNV workloads
 
         """
-        for cnv_wl in cnv_workloads:
+        # Iterating from end so that restored VMs are deleted before source
```
Comment (PrasadDesala) [inline tests/conftest.py:7094]: Is deleting the restored VMs before the source causing any problems? At least in PVCs, IIRC, I don't think there will be any issues. The parent PVC can be deleted even with the restored PVC in use

Comment (avd-sagare) [reply tests/conftest.py:7094]: I will take care of this in my next immediate PR.

File: tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py [hunk 5]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        # Stopping VM before taking snapshot of the VM PVC
+        vm_obj.stop()
+        # Taking Snapshot of PVC
+        pvc_obj = vm_obj.get_vm_pvc_obj()
+        snap_obj = snapshot_factory(pvc_obj)
+        # Restore the snapshot
+        res_snap_obj = snapshot_restore_factory(
+            snapshot_obj=snap_obj,
+            storageclass=vm_obj.sc_name,
+            size=vm_obj.pvc_size,
```
[the remark this answers is not in the document]
Comment (avd-sagare) [reply tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py:101]: Sure. I will remove this size in next PR.

File: tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py [hunk 6]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        )
+        # Create new VM using the restored PVC
+        res_vm_obj = cnv_workload(
+            volume_interface=constants.VM_VOLUME_PVC,
+            source_url=constants.CNV_FEDORA_SOURCE,
+            existing_pvc_obj=res_snap_obj,
+            namespace=vm_obj.namespace,
+        )[1]
+        # Write new file to VM
+        run_dd_io(vm_obj=res_vm_obj, file_path=file_paths[1], verify=True)
```
[the remark this answers is not in the document]
Comment (avd-sagare) [reply tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py:115]: Sure I will add this step after  validating the data integrity step.

File: tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py [hunk 7]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        # Validate data integrity of file written before taking snapshot
+        res_csum = cal_md5sum_vm(vm_obj=res_vm_obj, file_path=file_paths[0])
+        assert (
+            source_csum == res_csum
+        ), f"Failed: MD5 comparison between source {vm_obj.name} and cloned {res_vm_obj.name} VMs"
```
[the remark this answers is not in the document]
Comment (avd-sagare) [reply tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py:120]: I will change message in my PR.

File: tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py [hunk 8]
```diff
@@ -62,3 +62,60 @@ def test_vm_clone(self, cnv_workload, clone_vm_workload, setup_cnv):
...
+        res_vm_obj.stop()
```
[the remark this answers is not in the document]
Comment (avd-sagare) [reply tests/functional/workloads/cnv/test_vm_snapshot_cloning_ops.py:121]: I will correct step 7 to "Stopping all the VMs as part of this test before deleting"

**Review by shylesh — APPROVED** [review]
lgtm

File: tests/conftest.py (hunk 4 above)
Comment (shylesh) [reply tests/conftest.py:7094]: Based on avadhoot's comments, I will merge this PR as it's required urgently and low priority review comments will be handled in a followup PR

**Pull request merged**